File: java/src/NonBinaryTree.java
import java.util.ArrayList;
import java.util.List;

// Define a class for tree nodes
class TreeNode {
    int value;
    List<TreeNode> children;

    public TreeNode(int value) {
        this.value = value;
        this.children = new ArrayList<>();
    }

    public void addChild(TreeNode child) {
        children.add(child);
    }

    public void removeChild(TreeNode child) {
        children.remove(child);
    }
}

public class NonBinaryTree {
    TreeNode root;

    public NonBinaryTree(int value) {
        root = new TreeNode(value);
    }

    public static void main(String[] args) {
        NonBinaryTree tree = new NonBinaryTree(1);

        TreeNode childA = new TreeNode(2);
        TreeNode childB = new TreeNode(3);
        TreeNode childC = new TreeNode(4);

        tree.root.addChild(childA);
        tree.root.addChild(childB);
        tree.root.addChild(childC);

        TreeNode childAA = new TreeNode(5);
        TreeNode childAB = new TreeNode(6);

        childA.addChild(childAA);
        childA.addChild(childAB);

        // Remove childAB from childA
        childA.removeChild(childAB);
    }
}

/*
In this implementation, the TreeNode class represents a node in the non-binary tree. 
Each node has a value and a list of child nodes. 
The addChild and removeChild methods are provided to add and remove child nodes from a parent node.

The NonBinaryTree class represents the non-binary tree structure with a TreeNode as its root. 
The example provided in the main method demonstrates how to create a tree and add/remove child nodes from parent nodes.
 */

File: rust/src/non_binary_tree.rs
use std::cell::RefCell;
use std::rc::Rc;

// Define a struct for tree nodes
#[derive(Debug)]
struct TreeNode {
    value: i32,
    children: Vec<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    fn new(value: i32) -> TreeNode {
        TreeNode {
            value,
            children: Vec::new(),
        }
    }

    fn add_child(&mut self, child: Rc<RefCell<TreeNode>>) {
        self.children.push(child);
    }

    fn remove_child(&mut self, child_to_remove: &Rc<RefCell<TreeNode>>) {
        self.children
            .retain(|child| !Rc::ptr_eq(child, child_to_remove));
    }
}

struct NonBinaryTree {
    root: Rc<RefCell<TreeNode>>,
}

impl NonBinaryTree {
    fn new(value: i32) -> NonBinaryTree {
        NonBinaryTree {
            root: Rc::new(RefCell::new(TreeNode::new(value))),
        }
    }
}
